rfcs/text/0013_background_sessions.md

```diff
@@ -0,0 +1,498 @@
+
+   /**
+    * @param TODO: input arguments
+    * @returns a filtered list of BackgroundSessionAttributes objects. 
+    * @throws Throws an error in OSS.
+    */
+   list: (options: SavedObjectsFindOptions) => BackgroundSessionAttributes[]
+}
+```
+
+### Search Service Changes
+
+Since all search requests go through the `search_interceptor`, it will take on the responsibility for tracking the mapping between the request's hash and the async ID returned from Elasticsearch, as well as enriching the requests with any options required to restore an existing session.
+
+We're proposing the following changes to the `SearchInterceptor` API, and exposing them on the `search` service:
+
+```ts
+interface ISearchInterceptor {
+  /**
+   * Execute a search request. Uses synchronous execution in OSS and asynchronous execution in x-pack.
+   * If a new sessionId is provided, the `search_interceptor` will internally track the sessionId, a hash of `request.body.params` and the `asyncId` returned from Elasticsearch on the first request.
+   * If an existing session id is provided for a restore, the interceptor will define an option on the request, telling the server to use saved results, rather than re-running the Elasticsearch request. 
+   * @param request TODO define typing! The request contains an optional `id` parameter for the `sessionId`.
+   * @param options
+   * @returns
+   * @throws
+   */ 
+  search: (request: ISearchRequest<Payload=any>, options: TODO) => ISearchResponse<Payload=any>
+  
+  /**
+   * Calls `data.search.session.store()` with the provided sessionId and any mapped `async IDs`.
+   * @param sessionId
+   * @param name
+   * @param url TODO?
+   * @returns the session saved object 
+   * @throws Throws an error in OSS.
+   */
+  sendToBackground: (sessionId: string, name: string, url: string) => Promise<BackgroundSessionAttrbiutes>
+}
+
+```
+
+## Backend Services and Routes
+
+The server side's feature implementation builds on how Elasticsearch's `async_search` endpoint works. When making an
+initial new request to Elasticsearch, it returns a search ID that can be later used to retrieve the results.
+
+The server will then store that `request`, `sessionId`, and `searchId` in a mapping in memory, and periodically query
+for a saved object corresponding with that session. If the saved object is found, it will update the saved object to
+include this `request`/`searchId` combination, and remove it from memory. If, after a period of time (5 minutes?) the
+saved object has not been found, we will stop polling for that `sessionId` and remove the `request`/`searchId` from
+memory.  
+
+When the server receives a search request that has a `sessionId` and is marked as a `restore` request, the server will
+attempt to find the correct id within the saved object, and use it to retrieve the results previously saved.
+
+### New Session Service
+
+```ts
+interface ISessionService {
```

Comment:
shouldn't the server and client `new session service` have same interface ?

Reply:
++ IMO we should keep these the same unless there's some reason not to.